File: app/src/main/java/de/herrmann_engel/rbv/utils/SortCards.java
package de.herrmann_engel.rbv.utils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import de.herrmann_engel.rbv.Globals;
import de.herrmann_engel.rbv.db.DB_Card;

public class SortCards {

    private int compareCardsAlphabetical(String a, String b) {
        return a.compareToIgnoreCase(b);
    }

    public List<DB_Card> sortCards(List<DB_Card> list, int sort, Random random) {
        List<DB_Card> sortedList = new ArrayList<>(list);
        if (sort == Globals.SORT_ALPHABETICAL) {
            sortedList.sort((a, b) -> {
                int c0 = compareCardsAlphabetical(a.front, b.front);
                if (c0 == 0) {
                    return compareCardsAlphabetical(a.back, b.back);
                }
                return c0;
            });
            return sortedList;
        } else if (sort == Globals.SORT_RANDOM) {
            if (random == null) {
                Collections.shuffle(sortedList);
            } else {
                Collections.shuffle(sortedList, random);
            }
            return sortedList;
        }
        sortedList.sort((a, b) -> {
            int c0 = Integer.compare(a.known, b.known);
            if (c0 == 0) {
                return Long.compare(b.date, a.date);
            }
            return c0;
        });
        return sortedList;
    }

    public List<DB_Card> sortCards(List<DB_Card> list, int sort) {
        return sortCards(list, sort, null);
    }
}

File: app/src/main/java/de/herrmann_engel/rbv/utils/SortCards.kt
package de.herrmann_engel.rbv.utils

import de.herrmann_engel.rbv.Globals
import de.herrmann_engel.rbv.db.DB_Card_With_Meta

class SortCards {
    private fun compareCardsAlphabetical(a: String, b: String): Int {
        return a.compareTo(b, ignoreCase = true)
    }

    fun sortCards(list: MutableList<DB_Card_With_Meta>, sort: Int) {
        when (sort) {
            Globals.SORT_CARDS_ALPHABETICAL -> {
                list.sortWith(Comparator { a, b ->
                    return@Comparator compareCardsAlphabetical(
                            a.formattedFront ?: a.card.front,
                            b.formattedFront ?: b.card.front
                        ).let {
                            if (it == 0) {
                                compareCardsAlphabetical(
                                    a.formattedBack ?: a.card.back,
                                    b.formattedBack ?: b.card.back
                                )
                            } else it
                        }
                })
            }

            Globals.SORT_CARDS_RANDOM -> {
                list.shuffle()
            }

            Globals.SORT_CARDS_REPETITION -> {
                list.sortWith(compareByDescending<DB_Card_With_Meta> { it.card.lastRepetition }.thenByDescending { it.card.date })
            }

            Globals.SORT_CARDS_MIXED -> {
                list.forEach { c ->
                    c.mixedWeight = (if (c.card.known <= 5) {
                        (1 - c.card.known / 5.0) * 0.4
                    } else 0.0) + (if (c.card.countRepetitions > 0 && c.card.known <= c.card.countRepetitions) {
                        (1 - c.card.known / c.card.countRepetitions.toDouble()) * 0.3
                    } else 0.0) + Math.random() * 0.3
                }
                list.sortWith(compareByDescending { it.mixedWeight })
            }

            else -> {
                list.sortWith(compareBy<DB_Card_With_Meta> { it.card.known }.thenByDescending { it.card.date })
            }
        }
    }
}
